### PR TITLE
Leader Election issue #434

### PR DESCRIPTION
**Pull request by Invictus17**
Leader Election issue #434
https://github.com/kubernetes-client/python/issues/434

### CODE REVIEW

**Comment by Invictus17** [comment]
Hi @mbohlool @roycaihw , can you please review my PR?

**Comment by yliaog** [comment]
thanks for the PR.

could you please keep the file naming style consistent with the existing code? e.g. LeaderElection is better to switched to leaderelection

**Comment by Invictus17** [comment]
> thanks for the PR.
> 
> could you please keep the file naming style consistent with the existing code? e.g. LeaderElection is better to switched to leaderelection

Sure. I've pushed the changes.

File: leaderelection/leaderelection.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+            print(self.leaderElectionRecord['holderIdentity'], "failed to lead")
+            # the user is notified
+            print(self.getLatestLeader())
+            # A new leader has been established
+            self.follow(scheduler)
+        else:
+            # Make sure thread that runs onStoppedLeading callback is stopped
+            if self.OnStoppedLeadingThread:
+                self.OnStoppedLeadingThread.stop()
+                self.OnStoppedLeadingThread.join()
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:118]: `onStoppedLeading` is not used. Could you demonstrate how a user should react to `OnStoppedLeadingThread` in the example?

File: leaderelection/threadingwithexception.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import threading
+import ctypes
```
Comment (roycaihw) [inline leaderelection/threadingwithexception.py:2]: I'm not very excited if we have to add C compatibility to achieve this

File: leaderelection/leaderelection.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            # kill the OnLeading callback
+            self.leaderFunctionThread.stop()
+            self.leaderFunctionThread.join()
+
+            # create OnStoppedLeading thread
+            self.OnStoppedLeadingThread = threadingwithexception.thread(self.electionConfig.onStoppedLeading)
+            self.OnStoppedLeadingThread.start()
+
+            # Start to follow
+            self.follow(scheduler)
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:156]: In your implementation, a program can keep switching between being a leader and a follower. However reading the client-go implementation, when the leader fails to renew the lease, it [gives up and doesn't re-join as a follower](https://github.com/kubernetes/kubernetes/blob/f5334fad483856c41ceb1031036a96e96c32fcee/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L278-L279). Should we keep the client-go behavior and let the user decide what to do when a leader loses the lease?

File: leaderelection/leaderelection.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    def leaderElectorRecord(self, electionConfig):
+        return {"holderIdentity": str(electionConfig.lock.identity),
+                "leaseDurationSeconds": str(electionConfig.leaseDuration),
+                "acquireTime": str(datetime.datetime.now())}
+
+    # Returns latest leader
+    def getLatestLeader(self):
+        getStatus, getResponse = self.electionConfig.lock.Get(name=self.electionConfig.lock.name, namespace=self.electionConfig.lock.namespace)
+        if getStatus:
+            return "leader is " + str(ast.literal_eval(getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey])['holderIdentity'])
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:56]: why do we need `literal_eval`?

File: leaderelection/threadingwithexception.py (hunk 2 above)
Comment (Invictus17) [reply leaderelection/threadingwithexception.py:2]: I can look into a different way to kill threads.

File: leaderelection/leaderelection.py (hunk 3 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:156]: Sure. I can change the code to have a leader exit and close if failed to update lease.

File: leaderelection/leaderelection.py (hunk 1 above)
Comment (roycaihw) [reply leaderelection/leaderelection.py:118]: nvm, I misread the thread logic

File: leaderelection/leaderelection.py (hunk 4 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:56]: getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey] returns a string. 
literal_eval converts that string to a dictionary so that ['holderIdentity'] can be easily accessed.

**Comment by Invictus17** [comment]
Hi @roycaihw , Thanks for your review. Based on your comments I've pushed the suggested changes.

changes:  
1) Coding style - Updated variable and function names.  
2) Threading with traces - This implementation does not depend on ctypes.
3) Leader exits - The leader does not become a follower if it fails to update lease. It now exits after running onStoppedLeading().

File: leaderelection/threadingwithtrace.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# limitations under the License.
+
+import sys
+import trace
+import threading
+import time
+
+
+# citing source: https://www.geeksforgeeks.org/python-different-ways-to-kill-a-thread/
+class thread_with_trace(threading.Thread):
```
Comment (yliaog) [inline leaderelection/threadingwithtrace.py:22]: it would be simpler to use https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor than working with threads directly.

it is added in python 3, but i guess it is ok since python 2 is deprecated anyway, you don't have to support python 2.

File: leaderelection/leaderelection.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            print(self.leader_election_record['holderIdentity'], "failed to lead")
+            # the user is notified
+            print(self.getlatest_leader())
+            # A new leader has been established
+            self.follow(scheduler)
+        else:
+            # Make sure thread that runs onstopped_leading callback is stopped
+            if self.onstopped_leadingthread:
+                self.onstopped_leadingthread.stop()
+                self.onstopped_leadingthread.join()
```
Comment (yliaog) [inline leaderelection/leaderelection.py:133]: i don't think you need to 'stop', then 'join', i.e., wait for it to stop. 

it's better to implement a simpler leaderelection:
1. right at the start, try to acquire lease to be the leader
2. if not yet a leader, periodically check if it can acquire the lease to be leader
3. if it becomes the leader, call the hook on started_leading
4. try to maintain the leadership by renewing the lease
5. if fail to renew the lease, call the hook on_stopped_leading
6. done with the leaderelection, return

NOTE: the lifecycle of one leaderelection run has two possibilities:
a) it always blocks, waiting to be a leader, but never succeeds
b) it becomes a leader, lead for sometime, then stop leading and return

It is simpler because during one leaderelection run, it can be a leader at most *once*. if become a leader, then somehow lose the leadership, then the whole leaderelection returns. it leave to the caller of the leaderelection to decide what to do after that, the caller may exit completely, or the caller may choose to run another leaderelection.

File: leaderelection/leaderelection.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        get_status, get_response = self.election_config.lock.get(name=self.election_config.lock.name, namespace=self.election_config.lock.namespace)
+        if get_status:
+            return "leader is " + str(ast.literal_eval(get_response.metadata.annotations[self.election_config.lock.leader_electionrecord_annotationkey])['holderIdentity'])
+        else:
+            return "could not get the latest leader"
+
+    # Point of entry to Leader election
+    def run(self):
+        # Try to create/ acquire a lock
+        self.try_acquire_or_renew()
```
Comment (yliaog) [inline leaderelection/leaderelection.py:78]: better to check out the java implementation, https://github.com/kubernetes-client/java/blob/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java

It uses only three threads, or more precisely, three threadpools, but each pool has only a single thread.

File: leaderelection/resourcelock/configmaplock.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kubernetes.client.rest import ApiException
+from kubernetes import client, config
+from kubernetes.client.api_client import ApiClient
+
+
+class ConfigMapLock:
```
Comment (yliaog) [inline leaderelection/resourcelock/configmaplock.py:20]: maybe not now, but unittests need to be added before merging the PR

File: leaderelection/example.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+import uuid
+import leaderelection
+from kubernetes import client, config
+from resourcelock.configmaplock import ConfigMapLock
+import electionconfig
+from datetime import timedelta
+
+# Authenticate using config file
+config.load_kube_config(config_file=r"D:\Kubernetes open source - Python client\Go example\kubeconfig.txt")
```
[the remark this answers is not in the document]
Comment (Invictus17) [reply leaderelection/example.py]: sorry, I forgot to leave that out. Updated now.

File: leaderelection/leaderelection.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+    def try_acquire_or_renew(self):
+        # Check if lock is created
+        lock_status, lock_response = self.election_config.lock.get(self.election_config.lock.name, self.election_config.lock.namespace)
+
+        # create a default Election record for this candidate
+        self.leader_election_record = self.leaderelector_record(self.election_config)
```
[the remark this answers is not in the document]
Comment (Invictus17) [reply leaderelection/leaderelection.py:85]: Thank you for pointing it out. I'll update it in the following commit.

File: leaderelection/leaderelection.py (hunk 6 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:133]: Thank you for pointing this out. I should have removed this piece of code when I updated the code for a leader to not follow after losing leadership and exit. Other than that, the leader election logic is identical to what you've described. I'll also make sure that the program let's the user decide whether to exit or run for election again. I'll handle these changes in the next commit.

File: leaderelection/threadingwithexception.py (hunk 2 above)
Comment (cliffburdick) [reply leaderelection/threadingwithexception.py]: You should be able to use stop flags and let the client check for it.

File: leaderelection/leaderelection.py (hunk 4 above)
Comment (cliffburdick) [reply leaderelection/leaderelection.py]: > `getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey]` returns a string.
> `literal_eval `converts that string to a dictionary so that `['holderIdentity']` can be easily accessed.

Would json.loads() work or is it not JSON?

File: leaderelection/threadingwithexception.py (hunk 2 above)
Comment (Invictus17) [reply leaderelection/threadingwithexception.py]: > You should be able to use stop flags and let the client check for it.

Yes, that's similar to the latest update using traces(this commit is outdated now) but I also want to look into https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor as suggested by @yliaog in her review.

File: leaderelection/leaderelection.py (hunk 4 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py]: > > `getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey]` returns a string.
> > `literal_eval `converts that string to a dictionary so that `['holderIdentity']` can be easily accessed.
> 
> Would json.loads() work or is it not JSON?

No, it's not JSON.

File: leaderelection/leaderelection.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        self.oldelection_record = None
+
+        # islease_updated_follower is used by a follower candidate to mark if the leader has updated the lock object
+        self.islease_updated_follower = False
+        
+        # leaderfunction_thread contains the thread object for onstarted_leading
+        self.leaderfunction_thread = None
+        
+        # onstopped_leadingthread contains the thread object for onstopped_leading
+        self.onstopped_leadingthread = None
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:59]: on_stopped_leading should not be run in a separate thread

**Comment by Invictus17** [comment]
Hi @roycaihw @yliaog ,
Thank you for your review. I have updated my code based on your comments and discussions. The updates are:
1. Not killing threads and making the logic simpler, similar to the initial Java client.
2. OnStoppedLeading() is not run in a sub-thread.
3. Added a README.
4. Using logging and minor changes pointed out during review.

File: leaderelection/leaderelection.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            sys.exit("argument config not passed")
+
+        # The configuration set for this candidate
+        self.election_config = election_config
+        # The election record kept by this candidate
+        self.leader_election_record = None
+
+        # updatelease_schedulerId variable stores the scheduler object id for the update_lease schedule that is repeated
+        # every retry_period seconds by the leader
+        self.updatelease_schedulerId = None
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:36]: is this unused?

File: leaderelection/leaderelection.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+                # Start leading and call OnStartedLeading()
+                self.leaderfunction_thread = threading.Thread(target=self.election_config.onstarted_leading, daemon=True)
+                self.leaderfunction_thread.start()
+
+                # Schedule the function of updating the lock
+                scheduler_leader = sched.scheduler(time.time, time.sleep)
+                self.lead(scheduler_leader)
+
+    def transition_follower_to_leader(self):
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:120]: we don't need to distinguish `transition_follower_to_leader` and `lead`, same for `follow` and `try_acquire`. These complexity might be useful if we wanted to convert a leader back to a follower

File: leaderelection/leaderelection.py [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+            # compare timestamps
+            if self.oldelection_record['acquireTime'] != current_election_record['acquireTime']:
+                # Leader has updated lease continue following
+                self.islease_updated_follower = True
+
+            self.oldelection_record = current_election_record
+
+
+            # keep checking for lease updates every retry_period seconds
+            self.followerlease_checkscheduler = scheduler.enter(int(self.election_config.retry_period), 1, self.check_lease_updates, (scheduler,))
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:242]: previously I thought the scheduler was like the `wait.Until` method that client-go uses, where the scheduling and the real logic are isolated. I found this implementation (scheduling the real logic recursively) hard to follow. What's the benefit against using a simple while loop + sleep like the java client?

File: leaderelection/leaderelection.py (hunk 12 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:36]: No it is used:
https://github.com/Invictus17/python-base/blob/master/leaderelection/leaderelection.py#L138
https://github.com/Invictus17/python-base/blob/master/leaderelection/leaderelection.py#L182

File: leaderelection/leaderelection.py (hunk 14 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:242]: I don't think either has a benefit over the other. Would you suggest using a while + sleep?

File: leaderelection/leaderelection.py [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    def try_acquire_or_renew(self):
+        # Check if lock is created
+        lock_status, lock_response = self.election_config.lock.get(self.election_config.lock.name, self.election_config.lock.namespace)
+
+
+
+        # If a lock is already created with that name
+        if lock_status:
+            logging.info("{} is a follower".format(self.election_config.lock.identity))
+            scheduler_follower = sched.scheduler(time.time, time.sleep)
```
Comment (yliaog) [inline leaderelection/leaderelection.py:87]: there are a total of 4 of these: sched.scheduler(time.time, time.sleep) in the code. would one scheduler be sufficient? would it be better to create just one in the class __init__?

File: leaderelection/README.md [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+This example demonstrates how to use the leader election library.
+
+## Running
+Run the following command in multiple separate terminals preferably an odd number. Each running process uses a unique identifier displayed when it starts to run.
+When a program runs, it can either be a leader or a follower. The user will be prompted about the status of the candidate and transitions. 
+
+### Command to run
+```python example.py```
+
+Now kill the existing leader. You will see from the terminal outputs that one of the remaining two processes will be elected as the new leader.
```
[the remark this answers is not in the document]
Comment (Invictus17) [reply leaderelection/README.md:11]: Thanks for pointing it out. I'll update it.

File: leaderelection/leaderelection.py (hunk 15 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:87]: > there are a total of 4 of these: sched.scheduler(time.time, time.sleep) in the code. would one scheduler be sufficient? would it be better to create just one in the class **init**?

Yes, one would do, I'll update it. @roycaihw had a different opinion about using the sched module. Do you have similar thoughts about it?
https://github.com/kubernetes-client/python-base/pull/206#discussion_r471037636

Comment (yliaog) [reply leaderelection/leaderelection.py:87]: Yes, i agree with his comment. i think it's easier to understand to have the program structure like this:

1. try to acquire lease to be leader in main thread
2. if  get lease, then run lead in a separate thread
3. keep renewing lease in main thread
4. if failed renewing lease, then stop leading in main thread

**Comment by Invictus17** [comment]
Hi @roycaihw @yliaog ,
Thank you for your review. I have updated my code based on your comments and discussions. The updates are:
1. Switched from using sched module to a while + sleep approach.
2. Made the code consistent with the other clients.

File: leaderelection/leaderelection.py [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    def try_acquire_or_renew(self):
+        # Check if lock is created
+        lock_status, lock_response = self.election_config.lock.get(self.election_config.lock.name, self.election_config.lock.namespace)
+
+        # create a default Election record for this candidate
+        leader_election_record = self.leaderelector_record(self.election_config)
+        
+        # If a lock is already created with that name
+        if lock_status:
+            old_election_record = ast.literal_eval(lock_response.metadata.annotations[self.election_config.lock.leader_electionrecord_annotationkey])
```
[the remark this answers is not in the document]
Comment (yliaog) [reply leaderelection/leaderelection.py:108]: also ast.literal_eval could throw an exception, better to catch it

File: leaderelection/leaderelection.py [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+                logging.info("Leader has switched to {}".format(old_election_record['holderIdentity']))
+
+
+            if old_election_record != self.observed_record:
+                self.observed_record = old_election_record
+                self.observed_time_milliseconds = int(time.time()*1000)
+
+
+            # If This candidate is not the leader and lease duration is yet to finish
+            if str(self.election_config.lock.identity) != self.observed_record['holderIdentity'] and self.observed_time_milliseconds + self.election_config.lease_duration*1000 > int(time.time()*1000):
```
Comment (yliaog) [inline leaderelection/leaderelection.py:121]: self.observed_time_milliseconds is from this candidate, but it should be from the election record, no?

also self.observed_time_milliseconds is updated to the current time at line 117, so it will always be the current time, and the lease duration will never expire

File: leaderelection/leaderelection.py [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+                logging.info("yet to finish lease_duration, lease held by {} and has not expired".format(old_election_record['holderIdentity']))
+                return False
+
+            # If this candidate is the Leader
+            if str(self.election_config.lock.identity) == self.observed_record['holderIdentity']:
+                # Leader sets acquireTime
+                leader_election_record['acquireTime'] = self.observed_record['acquireTime']
```
Comment (yliaog) [inline leaderelection/leaderelection.py:128]: self.observed_record['acquireTime'] is not updated, it is still the old_election_record's acquireTime

File: leaderelection/leaderelection.py (hunk 18 above)
Comment (Invictus17) [reply leaderelection/leaderelection.py:121]: > self.observed_time_milliseconds is from this candidate, but it should be from the election record, no?
No, the java and go clients are also keeping a local record of 'observed_time_milliseconds' and not referring to the election record
https://github.com/kubernetes-client/java/blob/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L269-L315

 
> also self.observed_time_milliseconds is updated to the current time at line 117, so it will always be the current time, and the lease duration will never expire
It is only updated if a follower identifies that the lock object has been updated by a leader.
       if old_election_record != self.observed_record:
             # Update self.observed_time_milliseconds & self.observed_record

In case a leader fails to update the lock, self.observed_time_milliseconds will not be updated to current time & after a period of 'leaseDuration' a follower will try to update the lock. 
https://github.com/kubernetes-client/java/blob/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L267-L279

File: leaderelection/leaderelection.py [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+            old_election_record = ast.literal_eval(lock_response.metadata.annotations[self.election_config.lock.leader_electionrecord_annotationkey])
+
+            # Report transitions
+            if self.observed_record and self.observed_record['holderIdentity'] != old_election_record['holderIdentity']:
+                logging.info("Leader has switched to {}".format(old_election_record['holderIdentity']))
+
+
+            if old_election_record != self.observed_record:
+                self.observed_record = old_election_record
+                self.observed_time_milliseconds = int(time.time()*1000)
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:117]: capture the current time (e.g. `now = time.time()`) at the beginning of this function, and keep using `now`, instead of calling time.time() multiple times.

File: leaderelection/leaderelection.py (hunk 19 above)
Comment (roycaihw) [reply leaderelection/leaderelection.py:128]: this line is correct I think. It prevents leader_election_record['acquireTime'] to be "now"

File: leaderelection/leaderelection.py [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+                return False
+
+            # If this candidate is the Leader
+            if str(self.election_config.lock.identity) == self.observed_record['holderIdentity']:
+                # Leader sets acquireTime
+                leader_election_record['acquireTime'] = self.observed_record['acquireTime']
+
+            # Update object with latest election record
+            lock_response.metadata.annotations[self.election_config.lock.leader_electionrecord_annotationkey] = str(leader_election_record)
+            update_status, update_response = self.election_config.lock.update(self.election_config.lock.name, self.election_config.lock.namespace, lock_response)
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:132]: this is not generalized for different types of locks. Better to make leader_election_record a class or a dict, and pass it directly to the lock methods

Comment (roycaihw) [reply leaderelection/leaderelection.py:132]: the leaderelection logic doesn't need to know about `LeaderElectionRecordAnnotationKey`. It's a implementation detail for the lock

File: leaderelection/leaderelection.py (hunk 18 above)
Comment (yliaog) [reply leaderelection/leaderelection.py:121]: ok, the "> int(time.time()*1000):" in the above line is incorrect. time.time() would return the current time, what should be used instead is the time at the start of the function try_acquire_or_renew (https://github.com/kubernetes-client/java/blob/6a2a60ad2ad75fb127874797ea910b90e4a80651/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L236)

File: leaderelection/leaderelection.py (hunk 19 above)
Comment (yliaog) [reply leaderelection/leaderelection.py:128]: ok. i was thinking about setting the 'acquireTime' at the time when the lease is acquired, currently, the 'acquireTime' is set at the line: 104: leader_election_record = self.leaderelector_record(self.election_config)

it probably does not matter though, as there is no wait or loop inside the function try_acquire_or_renew

File: leaderelection/leaderelection.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+            return False
+
+        # If this candidate is the Leader
+        if str(self.election_config.lock.identity) == self.observed_record['holderIdentity']:
+            # Leader sets acquireTime
+            leader_election_record['acquireTime'] = self.observed_record['acquireTime']
+
+        return self.update_lock(lock_response, leader_election_record)
+
+    def update_lock(self, lock_response, leader_election_record):
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:168]: `lock_response` can be a field stored in the lock, since it was returned from `lock.get()` and never changed before we passed it back to `lock.update()`. See [client-go](https://github.com/kubernetes/client-go/blob/57f65489c11a4d4a978c93ab369c38f53156888a/tools/leaderelection/resourcelock/configmaplock.go#L41) as an example.

File: leaderelection/leaderelection.py [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+        # check for any key, value errors in the record
+        try:
+            old_election_record = ast.literal_eval(lock_record)
+            if (old_election_record['holderIdentity'] == '' or old_election_record['leaseDurationSeconds'] == ''
+                    or old_election_record['acquireTime'] == '' or old_election_record['renewTime'] == ''):
+                # try to update lock with proper annotation and election record
+                return self.update_lock(lock_response, leader_election_record)
+        except:
+            # try to update lock with proper annotation and election record
+            return self.update_lock(lock_response, leader_election_record)
```
Comment (roycaihw) [inline leaderelection/leaderelection.py:144]: I think it's better to assume `lock_record` is a class object, instead of a string here. Each lock type can have its own unmarshalling implementation (e.g. ConfigMap and Endpoints unmarshal their annotations, while Lease unmarshalls its spec). 

This will also allow the short-circuit logic here to be as simple as [client-java](https://github.com/kubernetes-client/java/blob/abea7f4dbf9bf523a4e94349129309da09e317e0/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L259-L264).

File: leaderelection/resourcelock/configmaplock.py [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    def create(self, name, namespace, election_record):
+        """
+        :param electionRecord: Annotation string
+        :param name: Name of the configmap object to be created
+        :param namespace: Namespace in which the configmap object is to be created
+        :return: 'True' if object is created else 'False' if failed
+        """
+        body = client.V1ConfigMap(
+            metadata={"name": name,
+                      "annotations": {self.leader_electionrecord_annotationkey: json.dumps(self.get_lock_dict(election_record))}})
```
Comment (roycaihw) [inline leaderelection/resourcelock/configmaplock.py:83]: `self.get_lock_dict` always updates `self.lock_record`. I think we don't want to update the lock record if the create fails. Same for `update()`.

**Review by roycaihw — COMMENTED** [review]
Thanks for generalizing the leader_election_record. Overall the implementation looks good to me. Please add a test.

File: leaderelection/resourcelock/configmaplock.py (hunk 24 above)
Comment (roycaihw) [reply leaderelection/resourcelock/configmaplock.py:83]: nvm, was thinking of `self.configmap_reference`

File: leaderelection/leaderelection.py [hunk 25]
```diff
@@ -101,82 +93,76 @@ def try_acquire_or_renew(self):
...
         # create a default Election record for this candidate
-        leader_election_record = self.leaderelector_record(self.election_config, now)
+        leader_election_record = LeaderElectionRecord(self.election_config.lock.identity,
+                                                     str(self.election_config.lease_duration), str(now), str(now))
 
         # A lock is not created with that name, try to create one
         if not lock_status:
-            if ast.literal_eval(lock_response.body)['code'] != 404:
-                logging.info("Error retrieving resource lock {} as {}".format(self.election_config.lock.name, lock_response.reason))
+            if json.loads(old_election_record.body)['code'] != 404:
```
Comment (yliaog) [inline leaderelection/leaderelection.py:105]: please avoid magic constant 404, instead use NOT_FOUND

File: leaderelection/leaderelection.py [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+        self.observed_time_milliseconds = 0
+
+    # Point of entry to Leader election
+    def run(self):
+        # Try to create/ acquire a lock
+        if self.acquire():
+            logging.info("{} successfully acquired lease".format(self.election_config.lock.identity))
+
+            # Start leading and call OnStartedLeading()
+            threading.Thread(target=self.election_config.onstarted_leading, daemon=True).start()
```
Comment (yliaog) [inline leaderelection/leaderelection.py:47]: please add some documentation to the class to describe the threading behavior. i.e. the thread is running until onstarted_leading returns if it does return, it's not safe to run this leader election more than once in a process, etc.

**Comment by Invictus17** [comment]
@yliaog @roycaihw In my latest push I've added the tests but the travis ci report is indicating a few failures. Any idea why some of these tests are failing?
test result: https://travis-ci.org/github/kubernetes-client/python-base/builds/751412124

**Comment by yliaog** [comment]
https://github.com/kubernetes-client/python-base/pull/222 fixed the configmap test failure

File: leaderelection/leaderelection_test.py [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+
+        def on_stopped_leading_A():
+            leadership_history.append("A stops leading")
+
+        config_A = electionconfig.Config(lock=mock_lock_A, lease_duration=2.5,
+                                         renew_deadline=2, retry_period=1.5, onstarted_leading=on_started_leading_A,
+                                         onstopped_leading=on_stopped_leading_A)
+
+        # Enter leader election
+        leaderelection.LeaderElection(config_A).run()
```
Comment (roycaihw) [inline leaderelection/leaderelection_test.py:87]: Since `run()` blocks until the leader election ends, B won't start until A hits `renew_count_max`. Could you make the two clients run in parallel to make sure the leader election handles concurrency properly?

File: leaderelection/leaderelection_test.py [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+        leaderelection.LeaderElection(config).run()
+
+        self.assert_history(election_history,
+                            ["create record",
+                             "try update record",
+                             "update record",
+                             "try update record",
+                             "update record",
+                             "try update record",
+                             "try update record"])
```
Comment (roycaihw) [inline leaderelection/leaderelection_test.py:172]: I had some trouble understanding what this test (`test_Leader_election_with_renew_deadline `) does. Could you add some comments to explain how it's related to `renew_deadline` and what's the expected behavior? Thanks

Comment (Invictus17) [reply leaderelection/leaderelection_test.py:172]: Thanks @roycaihw, sure I'll add some comments. The expected behavior is to check if the leader stops leading if it fails to update the lock after the renew_deadline.

File: leaderelection/leaderelection_test.py [hunk 29]
```diff
@@ -0,0 +1,267 @@
...
+    """Expected behavior: to check if the leader stops leading if it fails to update the lock within the renew_deadline
+    and stops leading after finally timing out. The difference between each try comes out to be approximately the sleep
+    time.
+    Example:
+    create record:  0s
+    on try update:  1.5s
+    on update:  zzz s
+    on try update:  3s
+    on update: zzz s 
+    on try update:  4.5s
```
Comment (roycaihw) [inline leaderelection/leaderelection_test.py:147]: Before this try, the timeout was set to be 4.5 + renew_deadline = 6.5s. After two failed tries (w/ sleep), the leader exited at 4.5+1.5*2=7.5s.

**Comment by roycaihw** [comment]
LGTM. Please squash the commits into reasonable parts.

File: leaderelection/leaderelection_test.py [hunk 30]
```diff
@@ -0,0 +1,267 @@
...
+
+class MockResourceLock:
+    def __init__(self, name, namespace, identity, on_create=None, on_update=None, on_change=None, on_try_update=None):
+        self.leader_record = []
+        self.renew_count = 0
+        self.renew_count_max = 4
+        self.name = name
+        self.namespace = namespace
+        self.identity = str(identity)
+        self.lock = threading.RLock()
```
Comment (roycaihw) [inline leaderelection/leaderelection_test.py:214]: is this lock per MockResourceLock? I'd expect it to be shared between all MockResourceLocks in one test

Comment (Invictus17) [reply leaderelection/leaderelection_test.py:214]: No, it should be shared. I'll update it. Thanks!

**Comment by roycaihw** [comment]
/lgtm
/approve

Thanks!
